File: crates/haver-vk/src/h264/annexb.rs
//! Annex B byte-stream helpers: split NAL units and classify them.

pub const NAL_SLICE: u8 = 1;
pub const NAL_IDR: u8 = 5;
pub const NAL_SPS: u8 = 7;
pub const NAL_PPS: u8 = 8;

/// One NAL unit: its type, `nal_ref_idc`, and its bytes starting at the
/// header byte (no start code).
#[derive(Debug, Clone, Copy)]
pub struct Nal<'a> {
    pub nal_type: u8,
    pub ref_idc: u8,
    pub data: &'a [u8],
}

/// Split an Annex B stream at its start codes.
pub fn nal_units(data: &[u8]) -> Vec<Nal<'_>> {
    let mut starts = Vec::new();
    let mut i = 0;
    while i + 3 <= data.len() {
        if data[i] == 0 && data[i + 1] == 0 && data[i + 2] == 1 {
            starts.push(i + 3);
            i += 3;
        } else {
            i += 1;
        }
    }
    starts
        .iter()
        .enumerate()
        .filter_map(|(n, &body)| {
            // Trailing zero bytes are never part of a NAL unit (the RBSP stop
            // bit and cabac_zero_words both end in a nonzero byte): they are
            // the leading zero of a 4-byte start code or trailing_zero_8bits.
            let mut end = starts.get(n + 1).map(|s| s - 3).unwrap_or(data.len());
            while end > body && data[end - 1] == 0 {
                end -= 1;
            }
            let unit = &data[body..end];
            unit.first().map(|h| Nal {
                nal_type: h & 0x1f,
                ref_idc: (h >> 5) & 3,
                data: unit,
            })
        })
        .collect()
}

pub fn has_start_code(data: &[u8]) -> bool {
    data.windows(3).any(|w| w == [0, 0, 1])
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn splits_units() {
        let s = [
            0, 0, 0, 1, 0x67, 1, 2, 0, 0, 1, 0x68, 3, 0, 0, 0, 1, 0x65, 9, 9,
        ];
        let units = nal_units(&s);
        let v: Vec<(u8, u8, usize)> = units
            .iter()
            .map(|n| (n.nal_type, n.ref_idc, n.data.len()))
            .collect();
        assert_eq!(v, vec![(NAL_SPS, 3, 3), (NAL_PPS, 3, 2), (NAL_IDR, 3, 3)]);
    }
}

File: macos/Sources/GliffVideo/AnnexB.swift
// Annex B byte-stream helpers: split an access unit into NAL units and
// rewrite it in the length-prefixed (AVCC) form VideoToolbox expects.
// Mirrors crates/gliff-vk/src/h264/annexb.rs.

import Foundation

public enum NalType: UInt8 {
    case slice = 1
    case idr = 5
    case sei = 6
    case sps = 7
    case pps = 8
    case accessUnitDelimiter = 9
}

/// One NAL unit, from its header byte on (no start code).
public struct Nal {
    public let data: Data

    public var type: UInt8 { data[data.startIndex] & 0x1f }
    public var isParameterSet: Bool {
        type == NalType.sps.rawValue || type == NalType.pps.rawValue
    }
}

/// Split an Annex B stream at its start codes. Trailing zero bytes are never
/// part of a NAL unit: they are the leading zero of a 4-byte start code or
/// trailing_zero_8bits.
public func nalUnits(_ data: Data) -> [Nal] {
    let bytes = [UInt8](data)
    var starts: [Int] = []
    var i = 0
    while i + 3 <= bytes.count {
        if bytes[i] == 0 && bytes[i + 1] == 0 && bytes[i + 2] == 1 {
            starts.append(i + 3)
            i += 3
        } else {
            i += 1
        }
    }
    var units: [Nal] = []
    for (n, body) in starts.enumerated() {
        var end = n + 1 < starts.count ? starts[n + 1] - 3 : bytes.count
        while end > body && bytes[end - 1] == 0 {
            end -= 1
        }
        if end > body {
            units.append(Nal(data: Data(bytes[body..<end])))
        }
    }
    return units
}

/// An access unit split into what VideoToolbox needs: the parameter sets, if
/// the unit carries them, and the remaining NAL units as AVCC (each prefixed
/// with its 4-byte big-endian length).
public struct AccessUnit {
    public var sps: Data?
    public var pps: Data?
    public var avcc = Data()

    public init(annexB: Data) {
        for nal in nalUnits(annexB) {
            switch nal.type {
            case NalType.sps.rawValue: sps = nal.data
            case NalType.pps.rawValue: pps = nal.data
            case NalType.accessUnitDelimiter.rawValue: break
            default:
                var length = UInt32(nal.data.count).bigEndian
                avcc.append(Data(bytes: &length, count: 4))
                avcc.append(nal.data)
            }
        }
    }
}
